### PR TITLE
Roles for controllers and test cases added

### PR DESCRIPTION
**Pull request by lalithareddy007**
#Added @PreAuthorize annotation to define roles, like "hasRole('ROLE_ADMIN') or hasRole('ROLE_STAFF')". 
#Added annotations to all test cases according to the roles of respective endpoints like @WithMockAdminStaff,@WithMockStaffStudent,  @WithMockStaff, @WithMockAdmin, @WithMockUser.
#Added negative test cases for roles by using .isForbidden() method.

### CODE REVIEW

File: src/main/java/com/numpyninja/lms/controller/AssignmentController.java [hunk 1]
```diff
@@ -39,7 +40,7 @@ public class AssignmentController {
 	
 	//create an assignment
 	@PostMapping("")
-	@RolesAllowed({"ROLE_ADMIN"})
+	@PreAuthorize("hasRole('ROLE_STAFF')")
```
Comment (gkav01) [inline src/main/java/com/numpyninja/lms/controller/AssignmentController.java:43]: Please add admin role also

File: src/main/java/com/numpyninja/lms/controller/AssignmentController.java [hunk 2]
```diff
@@ -48,7 +49,7 @@ public ResponseEntity<AssignmentDto> createAssignment(@Valid @RequestBody Assign
 	
 	//update an assignment
 	@PutMapping("/{id}")
-	@RolesAllowed({"ROLE_ADMIN"})
+	@PreAuthorize("hasRole('ROLE_STAFF')")
```
Comment (gkav01) [inline src/main/java/com/numpyninja/lms/controller/AssignmentController.java:52]: Please add admin role too, I have updated the excel sheet

File: src/main/java/com/numpyninja/lms/controller/AssignmentController.java [hunk 3]
```diff
@@ -57,7 +58,7 @@ public ResponseEntity<AssignmentDto> updateAssignment(@Valid @RequestBody Assign
 	
 	//delete an assignment
 	@DeleteMapping("/{id}")
-	@RolesAllowed({"ROLE_ADMIN"})
+	@PreAuthorize("hasRole('ROLE_STAFF')")
```
Comment (gkav01) [inline src/main/java/com/numpyninja/lms/controller/AssignmentController.java:61]: Same here, pls allow admin role as well.

File: src/test/java/com/numpyninja/lms/controller/AssignmentControllerTest.java [hunk 4]
```diff
@@ -116,10 +133,29 @@ void testUpdateAssignment() throws Exception {
                 .andExpect(jsonPath("$.assignmentId", is(updatedAssignmentDto.getAssignmentId()), Long.class))
                 .andExpect(jsonPath("$.assignmentName", is(updatedAssignmentDto.getAssignmentName())));
     }
+    @DisplayName("test for updating an assignment")
+    @Test
+    @WithMockAdmin
+    void testUpdateAssignmentByAdmin() throws Exception {
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/AssignmentControllerTest.java:139]: Please allow Admin as well. Please make test changes according to proposed changes in controller.

File: src/test/java/com/numpyninja/lms/controller/AssignmentControllerTest.java [hunk 5]
```diff
@@ -132,8 +168,23 @@ void testDeleteAssignment() throws Exception {
         response.andExpect(status().isOk())
                 .andDo(print());
     }
+    @DisplayName("test for deleting an assignment")
+    @Test
+    @WithMockAdmin
+    void testDeleteAssignmentByAdmin() throws Exception {
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/AssignmentControllerTest.java:174]: same as above

File: src/test/java/com/numpyninja/lms/controller/AssignmentSubmitControllerTest.java [hunk 6]
```diff
@@ -285,6 +323,20 @@ public void testGradeSubmissions() {
         verify(assignmentSubmitService).gradeAssignmentSubmission(any(AssignmentSubmitDTO.class), eq(submissionId));
 
     }
+
+    @Test
+    @SneakyThrows
+    @DisplayName("Test to grade submissions")
+    @WithMockAdmin
+    public void testGradeSubmissionsByAdmin() {
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/AssignmentSubmitControllerTest.java:331]: Pls allow admin role

File: src/test/java/com/numpyninja/lms/controller/AssignmentSubmitControllerTest.java [hunk 7]
```diff
@@ -267,8 +287,26 @@ public void testResubmitAssignment() {
 
     @Test
     @SneakyThrows
-    @DisplayName("Test to grade submissions")
+    @DisplayName("Test for  ResubmitAssignment")
     @WithMockAdminStaff
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/AssignmentSubmitControllerTest.java:291]: This indicates a single  user with Admin and staff roles. We also need to check for user with single role, either admin or staff

File: src/test/java/com/numpyninja/lms/controller/GoogleCalendarControllerTest.java [hunk 8]
```diff
@@ -113,6 +120,19 @@ void createCalendarEvent() {
 			.andExpect(MockMvcResultMatchers.jsonPath("$.eventStatus", equalTo(mockEventResponse.getEventStatus())));
 		 verify(calendarService).createEventUsingServiceAcc(Mockito.any(GCalendarEventRequestDTO.class));
 	}
+	@Test
+	@SneakyThrows
+	@DisplayName("GCalendarTest: CreateCalendarEvent")
+	@WithMockStaffStudent
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/GoogleCalendarControllerTest.java:126]: Same as above, create separate user and staff

File: src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java [hunk 9]
```diff
@@ -68,13 +68,22 @@ public static void setData() {
 
 
     @Test
+    @WithMockAdminStaff
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java:71]: Same as above

File: src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java [hunk 10]
```diff
@@ -154,7 +163,8 @@ public void givenBatch_whenCreateBatch_thenReturnSavedBatch() throws Exception {
                 .andExpect(jsonPath("$.batchName", is(batchDTO3.getBatchName())));
     }
 
-    @Test @WithMockStaffStudent
+    @Test
+    @WithMockStaffStudent
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java:167]: Same as above

File: src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java [hunk 11]
```diff
@@ -185,7 +195,7 @@ public void givenUpdatedBatch_whenUpdateBatch_thenReturnUpdateBatchObject_Role_A
                 .andExpect(jsonPath("batchDescription", is(updateDetailDTO.getBatchDescription())));
     }
 
-    @Test @WithMockStudent   // 403 forbidden
+    @Test @WithMockStaffStudent   // 403 forbidden
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java:198]: Same as above

File: src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java [hunk 12]
```diff
@@ -215,7 +225,7 @@ public void givenBatchId_whenDeleteBatch_thenReturn200() throws Exception {
                 .andDo(print());
     }
 
-    @Test @WithMockStaff
+    @Test @WithMockStaffStudent
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/ProgBatchControllerTest.java:228]: Same as above

File: src/test/java/com/numpyninja/lms/controller/ProgramControllerTest.java [hunk 13]
```diff
@@ -183,8 +184,8 @@ public void testCreateProgram() {
     @DisplayName("Test for Creating a Program")
     @Test
     @SneakyThrows
-    @WithMockStaff
-    public void testCreateProgramByStaff() {
+    @WithMockStaffStudent
```
Comment (gkav01) [inline src/test/java/com/numpyninja/lms/controller/ProgramControllerTest.java:187]: Same, pls check in all places where two roles are used to mack. two roles means one user having both roles. 
We als o need to check for one user having one role

**Comment by lalithareddy007** [comment]
Added separate test cases for each role.
Did change according to Excel sheet for assignment and assignment submission controllers

**Pull request merged**